### ISSUE TITLE
Enhancement: M1 support

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#613]
This ticket tracks the development of an M1 version.

**Comment by mpogue2** [comment]
So far, I have compiled (with some mods to the build process and the source code, which I'll check in at some point):

- SD 39.39 - Sept 28 2021 (latest version!) ==> libsdlib.a
- taglib 1.12 - Feb 16 2021 (latest version!) ==> libtaglib.a

**Comment by mpogue2** [comment]
quazip will compile, but it will not link on M1, because it uses the Qt5.15 compatibility lib, which is X86 only.
We'll need to find an alternative for this.

libbass is not available for M1 Mac.  I've proposed alternatives in a separate ticket.

mp3gain is deprecated.  I'm going to remove this functionality, so this library is not needed right now.

**Comment by mpogue2** [comment]
TODO (later):
pocketsphinx: voice recognition for Plus calls for SD
qpdfjs: PDF reader for the Resources tab
html-tidy: fixes up HTML for the Lyrics editor

**Comment by mpogue2** [comment]
Note that we should do some of these for the Intel version anyway, like moving to SD latest and taglib latest, moving to Qt6.2 (which involves getting rid of QRegExp, among other things).  

I'll try to make these changes in a backward compatible way, so that the Intel build continues on just fine.

**Comment by mpogue2** [comment]
#582 has updated SD.  This compiles on M1, but I have not yet tested it running, since the UI is not yet up on M1.

**Comment by mpogue2** [comment]
Making some progress here.   Everything compiles and runs, and does not crash.  
I have stubs in for a lot of stuff, like music loading/playback.  But, the GUI seems to work fine.  
![Screen Shot 2021-10-31 at 3 51 17 PM](https://user-images.githubusercontent.com/6316003/139603875-13a959c4-6189-4e4d-9c84-7e8874d4f096.png)

**Comment by mpogue2** [comment]
Sd works fine on Native M1 Mac, and this is the latest version of SD, too!

![Screen Shot 2021-10-31 at 3 51 41 PM](https://user-images.githubusercontent.com/6316003/139603894-8917fd89-942a-4693-b32f-7d0ff0e6f1ef.png)

**Comment by mpogue2** [comment]
Problems:  The Reference tab generates a lot of runtime errors, like these:

```
//    [8121:17667:1031/151459.045273:ERROR:channel_mac.cc(499)] mach_msg receive: (ipc/rcv) msg too large (0x10004004)
//    [8119:259:1031/151459.066135:ERROR:network_service_instance_impl.cc(330)] Network service crashed, restarting service.

```
So, the Reference tab is turned off right now, too.  It seems that other people have seen this problem with Qt6.2 on M1 MacBook, but there are no answers or workarounds that I could find yet.

**Comment by mpogue2** [comment]
I also got rid of some more warnings while I was at it:
- around TKeySequence
- around use of .SF NS Text taking an extra 400ms at startup time (changed those usages to Arial)

**Comment by mpogue2** [comment]
FYI: a complete from-scratch clean build on Mac M1 currently takes 52 seconds.  All 8 cores are 100% busy (as per menumeters).  

Nice.

This is also not all the code (because not everything is in the build yet), but it's basically sd + taglib + squaredesk.
I decided to get the M1 MacBook instead of the M1 Pro/Max, because I think 8 cores is enough, and I don't have a huge need for more GPU cores.

**Comment by mpogue2** [comment]
Lyrics editor seems to work, although I have all the libtidy stuff turned OFF right now (need to recompile the libtidy library for M1).  That means the some weird stuff might happen when editing existing HTML that hasn't already been "tidied".

![Screen Shot 2021-10-31 at 4 09 04 PM](https://user-images.githubusercontent.com/6316003/139604317-897a6dae-e662-4564-8d84-e11dd8e19da7.png)

**Comment by mpogue2** [comment]
As of 99f39a775eef8005757aee5faf51daefdf96b9a7, the source code compiles and runs on both X86 and M1 MacBooks.

The .pro file needs to be modified still, to turn on the M1 stuff, and turn off the X86 stuff, but overall I'm happy that work can proceed on the M1 side.

**Comment by mpogue2** [comment]
BUG: Debug version works fine, with no apparent loss of CPU.  Release version crashes in SD somewhere in initialization. Should fix this when I get a chance.

**Comment by mpogue2** [comment]
By the way, the Dance Programs tab works fine, too.

**Comment by mpogue2** [comment]
Removed PocketSphinx and voice input: bfaa65c541374e251eccec6876b690104fadc3c9
Removed ReplayGain: bd5801a7968dd1abba2e572f2253cdd4ff3e31c3
Remove Compressor: a50ce149a90ce9bff28f4460dd383984b77d7adc
Removed Timers tab: 3d3fafae90e135ce06987988bae7086fa4f5fc39
References tab invisible for now, until Qt bug is fixed: 1833ca0eebac64e639be14ad484cf50a00343cae

**WORKING:**
- MUSIC PLAYER: song filters, load, playback, loops, force mono, EQ (b/m/t), auto BPM detect, pitch/tempo change, mix, VU meter, Volume, Clock, tags, sound FX
- LYRICS: editing, auto-scrolling, setting loop points (they seem more precise now, too!)
- SD: everything (no voice input), including can say "HEADS Square Thru 4" to start (no HEADS START required)
- DANCE PROGRAMS: everything
- PREFERENCES: everything (no compressor, no replay gain, intelligibility boost not hooked up yet)

**TODO:**

Short term:
- Hook up Intelligibility Boost
- Combine flexible_audio and AudioDecoder classes (cleanup)
- Check to make sure that PackageIt.command works for M1
- Verify that same source code compiles/runs on X86 Mac
- Check to make sure that PackageIt.command still works for X86 

Longer term:
- Get References tab working (later, after Qt fixes their bug)
- Figure out how to make a unified app that handles both architectures: X86 and M1
- Try again to make app signing work (in an automated way, too, perhaps adding it to the PackageIt.command script would be ideal)

**Comment by mpogue2** [comment]
According to the QT bug (QTBUG-98487), the References tab should be possible to bring back now.
The last test I did for this was on Qt 6.2.1, so I will upgrade to Qt 6.2.2 and will try again.

From the QTBUG:
"I have update to Qt 6.2.2 and QtCreator 6.0 -> seems to work now ! I continue to try and let you know if the problem appears again"

**Comment by mpogue2** [comment]
References tab still isn't ready to go back in yet.
But, everything else works pretty well on M1.  It's fast and reliable so far.
The flexible_audio branch has all the M1 stuff, PLUS it has a lot of bug fixes.  So, when that's merged back in, goodness will reign.

**Comment by mpogue2** [comment]
Everything is in now that's going to be in.  FlexibleAudio branch has been merged to Master by ee843859129d8b5074cb5cd00d517b9669ec8a41, so closing this one as DONE.
